python/cugraph-pyg/pyproject.toml

```diff
@@ -31,6 +31,7 @@ dependencies = [
     "numba>=0.57",
     "numpy>=1.23,<2.0a0",
     "pylibcugraphops==24.6.*",
+    "tensordict>=0.1.2,<0.3.1a0",
```

Comment:
I didn't catch this in my first pass, but this does not really enforce an upper bound for pytorch:
```
tensordict 0.3.0: torch>=2.1.0
tensordict 0.3.1: torch>=2.2.1
tensordict 0.3.2: torch==2.2.2
```
In addition, this line now introduces pytorch as a hard dependency for cugraph-pyg, and it will pull pytorch from PyPI, as opposed to using its own index-url (as mentioned in https://pytorch.org/get-started/locally/). I am not clear about the consequence such as if it might pull a CPU only version from PyPI.

If you look at torch_geometric's pyproject.toml, they don't even include pytorch as a hard dependency. One alternative is to run version check at import time from cugraph-pyg

Reply:
Can't you just manually install the GPU version (as we do now)?

Reply:
Yes, of course you always manually install the correct GPU version to override whatever is installed by `pip install cugraph-pyg`

Reply:
I know it's not ideal but otherwise we would be requiring users to manually install `tensordict` which would be unexpected.  Right now I assume most of them do a manual install of `torch` anyways.

Reply:
I was just about to post something similar after reading the @alexbarghi-nv 's explanation in https://github.com/rapidsai/cugraph/pull/4464#issuecomment-2153126553 (thanks for that!). I saw these same version constraints looking through `tensordict` releases (e.g. https://github.com/rapidsai/cugraph/pull/4464#issuecomment-2153126553).

If the core issue is "this project's CI setup is incompatible with testing against PyTorch `>= 2.3.0`", then yeah similar to what was suggested in that comment, I think a better immediate fix would be:

* put a pin of the form `pytorch <2.3.0` in the testing dependencies
* don't modify anything about the project's `tensordict` dependency

That should lead the `pip` and `conda` solvers to avoid `tensordict` 0.4.0, because it's requirement on `torch>=2.3.1` wouldn't be satisfiable.

And it'd still make it possible for `cugraph` use to install `cugraph==24.6.0` alongside a newer `nccl` and newer `pytorch`.

`tensordict`'s 0.4.0 release happened to be the thing that introduced `pytorch >=2.3.0` here, but some other dependency could do that in the future. Putting a ceiling in CI on `pytorch` is the most reliable way to prevent that from happening again... and hopefully for the next release you can work out a way to test against `pytorch >=2.3.0` here in CI (maybe by upgrading beyond the `cupy` / `nccl` versions that are leading to the issue you've observed).

Reply:
Ok, I'll make that change.

Reply:
We still want the tensordict pin change, though.  Otherwise we are locking out users on PyTorch 2.2 and 2.1 which we do not want.

Reply:
I don't think so?

The pin is currently `tensordict>=0.1.2`.

https://github.com/rapidsai/cugraph/blob/286591ea8361d85a129f2a98025d981982d67e92/conda/recipes/cugraph-pyg/meta.yaml#L37

That's sufficiently old to be installed alongside `pytorch==2.1.0`.

I just tried this:

```shell
conda create --name pytorch-test--dry-run \
  -c conda-forge \
  -c pytorch \
    'tensordict>=0.1.2' \
    'pytorch==2.1.0'
```

And got a successful solution like this:

```text
libtorch           conda-forge/linux-64::libtorch-2.1.0-cuda118_h7595d50_303
...
pytorch            conda-forge/linux-64::pytorch-2.1.0-cuda118_py312hffed1d1_303
...
tensordict         conda-forge/noarch::tensordict-0.1.2-pyh6074d0b_0
```

<details><summary>output of 'conda info' (click me)</summary>

```text
     active environment : None
       user config file : /home/nfs/jlamb/.condarc
 populated config files : /raid/jlamb/miniforge/.condarc
                          /home/nfs/jlamb/.condarc
          conda version : 24.3.0
    conda-build version : not installed
         python version : 3.10.12.final.0
                 solver : libmamba (default)
       virtual packages : __archspec=1=broadwell
                          __conda=24.3.0=0
                          __cuda=12.2=0
                          __glibc=2.31=0
                          __linux=5.4.0=0
                          __unix=0=0
       base environment : /raid/jlamb/miniforge  (writable)
      conda av data dir : /raid/jlamb/miniforge/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /raid/jlamb/miniforge/pkgs
                          /home/nfs/jlamb/.conda/pkgs
       envs directories : /raid/jlamb/miniforge/envs
                          /home/nfs/jlamb/.conda/envs
               platform : linux-64
             user-agent : conda/24.3.0 requests/2.31.0 CPython/3.10.12 Linux/5.4.0-182-generic ubuntu/20.04.6 glibc/2.31 solver/libmamba conda-libmamba-solver/24.1.0 libmambapy/1.5.8
                UID:GID : 10349:10004
             netrc file : None
           offline mode : False
```

</details>

Reply:
Sorry, you're right.  I'll revert it.

Reply:
the latest changes look good to me! Thanks for hearing me out.

Reply:
Sure, no problem